File: backend/app/routers/legislation.py
import os
import subprocess
from fastapi import APIRouter, HTTPException, Query, UploadFile, File, Form, Depends
from typing import List, Optional, Dict, Any
import asyncio
from app.services.legislation_service import LegislationService
from app.services.extractor_service import ExtractorService
from app.services.legislation_crawler import LegislationCrawlerService
from app.lib.auth import get_current_user, require_roles
from app.config import get_settings
settings = get_settings()

MEVZUAT_DIR = settings.MEVZUAT_DIR

router = APIRouter(tags=["legislation"])


def _is_admin_like(current_user: Dict[str, Any]) -> bool:
    role = (current_user.get("role") or "user").strip().lower()
    return role in {"admin", "moderator"}

@router.post("/fetch-external")
async def fetch_external_legislation(data: dict, current_user: Dict[str, Any] = Depends(get_current_user)):
    """Fetches legislation metadata from an external source like mevzuat.gov.tr"""
    url = data.get("url")
    if not url:
        raise HTTPException(status_code=400, detail="URL belirtilmedi.")
    
    result = await LegislationCrawlerService.fetch_from_mevzuat_gov_tr(url)
    if not result:
        raise HTTPException(status_code=400, detail="Mevzuat bilgileri çekilemedi. Lütfen URL'yi kontrol edin.")
    return result

@router.post("/open-folder")
async def open_legislation_folder(
    category: Optional[str] = Query(None),
    doc_type: Optional[str] = Query(None),
    current_user: Dict[str, Any] = Depends(require_roles("admin", "moderator")),
):
    """Opens a Windows Explorer folder for the specified category/type or the root directory."""
    try:
        path = MEVZUAT_DIR
        if category and category != "Tümü":
            if doc_type:
                path = os.path.join(MEVZUAT_DIR, category, doc_type)
            else:
                path = os.path.join(MEVZUAT_DIR, category)
            
        if not os.path.exists(path):
            os.makedirs(path, exist_ok=True)
            
        # os.startfile is non-blocking on Windows
        os.startfile(os.path.abspath(path))
        return {"status": "success", "path": path}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

from app.schemas.legislation import LegislationCreate, LegislationUpdate, LegislationResponse

@router.post("/upload")
async def upload_legislation_file(
    file: UploadFile = File(...),
    category: str = Form(...),
    doc_type: str = Form(""),
    uid: Optional[str] = Form(None),
    is_public: bool = Form(True),
    current_user: Dict[str, Any] = Depends(get_current_user),
):
    """Uploads a file to Mevzuat. Can be shared (Genel) or Personal (Kisisel)."""
    try:
        # Validate extension
        ext = file.filename.split('.')[-1].lower()
        if ext not in ['pdf', 'doc', 'docx', 'jpg', 'jpeg', 'png']:
            raise HTTPException(status_code=400, detail="Desteklenmeyen dosya formatı. (PDF, Word veya Resim yükleyin)")
            
        is_admin = _is_admin_like(current_user)
        caller_uid = current_user.get("uid")

        if not is_admin and is_public:
            raise HTTPException(status_code=403, detail="Genel mevzuat yükleme yalnızca yönetici/moderatör hesaplarına açıktır.")

        if not is_admin and uid and uid != caller_uid:
            raise HTTPException(status_code=403, detail="Başka kullanıcı adına mevzuat yükleyemezsiniz.")

        target_uid = uid
        if not is_public and not is_admin:
            target_uid = caller_uid

        file_url = await LegislationService.save_legislation_file(file, category, doc_type, target_uid, is_public)
        return {"file_url": file_url}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.post("/extract-text")
async def extract_text(file: UploadFile = File(...), current_user: Dict[str, Any] = Depends(get_current_user)):
    """Extracts text from an uploaded PDF or Word document."""
    try:
        text = await ExtractorService.extract_text(file)
        return {"text": text}
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Metin ayıklama sırasında hata oluştu: {str(e)}")

@router.get("/", response_model=List[LegislationResponse])
async def get_legislations(
    uid: Optional[str] = Query(None),
    category: Optional[str] = Query(None, description="Mevzuat kategorisi (Genel, Personel, vb.)"),
    current_user: Dict[str, Any] = Depends(get_current_user),
):
    try:
        is_admin = _is_admin_like(current_user)
        caller_uid = current_user.get("uid")
        if uid and not is_admin and uid != caller_uid:
            raise HTTPException(status_code=403, detail="Başka kullanıcının özel mevzuatına erişemezsiniz.")
        return await LegislationService.get_legislations(uid, category, is_admin)
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/", response_model=LegislationResponse)
async def create_legislation(legislation: LegislationCreate, current_user: Dict[str, Any] = Depends(get_current_user)):
    try:
        is_admin = _is_admin_like(current_user)
        caller_uid = current_user.get("uid")

        if not is_admin and legislation.owner_id and legislation.owner_id != caller_uid:
            raise HTTPException(status_code=403, detail="Başka kullanıcı adına mevzuat oluşturamazsınız.")

        payload = legislation
        if not is_admin:
            payload = legislation.copy(update={"owner_id": caller_uid if not legislation.is_public else None})

        return await LegislationService.create_legislation(payload, is_admin)
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.post("/{leg_id}/approve")
async def approve_legislation(
    leg_id: str,
    admin_name: str = Query(...),
    current_user: Dict[str, Any] = Depends(require_roles("admin", "moderator")),
):
    success = await LegislationService.approve_legislation(leg_id, admin_name)
    if not success:
        raise HTTPException(status_code=404, detail="Mevzuat bulunamadı.")
    return {"status": "success"}

@router.post("/{leg_id}/reject")
async def reject_legislation(
    leg_id: str,
    current_user: Dict[str, Any] = Depends(require_roles("admin", "moderator")),
):
    success = await LegislationService.reject_legislation(leg_id)
    if not success:
        raise HTTPException(status_code=404, detail="Mevzuat bulunamadı.")
    return {"status": "success"}

@router.patch("/{leg_id}", response_model=LegislationResponse)
async def update_legislation(
    leg_id: str,
    leg_update: LegislationUpdate,
    current_user: Dict[str, Any] = Depends(require_roles("admin", "moderator")),
):
    updated = await LegislationService.update_legislation(leg_id, leg_update)
    if not updated:
        raise HTTPException(status_code=404, detail="Mevzuat bulunamadı.")
    return updated

@router.delete("/{leg_id}")
async def delete_legislation(
    leg_id: str,
    current_user: Dict[str, Any] = Depends(require_roles("admin", "moderator")),
):
    success = await LegislationService.delete_legislation(leg_id)
    if not success:
        raise HTTPException(status_code=404, detail="Mevzuat silinemedi.")
    return {"status": "success", "message": "Mevzuat silindi."}

@router.post("/{leg_id}/promote", response_model=LegislationResponse)
async def promote_legislation(
    leg_id: str,
    user_name: str = Query(...),
    current_user: Dict[str, Any] = Depends(require_roles("admin", "moderator")),
):
    """Promotes a private legislation into the public shared library."""
    promoted = await LegislationService.promote_to_public(leg_id, user_name)
    if not promoted:
        raise HTTPException(status_code=404, detail="Mevzuat bulunamadı.")
    return promoted

File: backups/AI_Auto_Backup_20260523_165115/frontend/src/lib/api/legislation.ts
import { API_URL as API_BASE_URL } from "../config";
import { fetchWithTimeout } from "./utils";

export interface Legislation {
    id: string;
    title: string;
    category: string;
    doc_type: string;
    summary?: string;
    content?: string;
    tags?: string[];
    official_gazette_info?: string;
    document_url?: string;
    local_path?: string;
    is_approved: boolean;
    approved_by?: string;
    approved_at?: string;
    owner_id?: string;
    created_by_name?: string;
    last_updated_by_name?: string;
    is_public: boolean;
    is_pinned: boolean;
    is_archived: boolean;
    created_at: string;
}

export const promoteToPublic = async (id: string, userName: string): Promise<Legislation> => {
    const res = await fetchWithTimeout(`${API_BASE_URL}/legislation/${id}/promote?user_name=${encodeURIComponent(userName)}`, {
        method: "POST"
    });
    return res.json();
};

export interface LegislationCreate {
    owner_id?: string;
    is_public: boolean;
    title: string;
    category: string;
    summary?: string;
    content: string;
    tags: string[];
    document_url?: string;
    official_gazette_info?: string;
    is_pinned: boolean;
}

export async function fetchLegislations(uid?: string, category?: string, isAdmin: boolean = false): Promise<Legislation[]> {
    const params = new URLSearchParams();
    if (uid) params.append("uid", uid);
    if (category && category !== 'All' && category !== 'Tümü') {
        params.append("category", category);
    }
    if (isAdmin) params.append("is_admin", "true");
    
    const url = `${API_BASE_URL}/legislation/${params.toString() ? '?' + params.toString() : ''}`;
    
    const response = await fetchWithTimeout(url);
    if (!response.ok) {
        throw new Error("Mevzuat yüklenemedi.");
    }
    return response.json();
}

export async function uploadLegislationFile(
    file: File, 
    category: string, 
    doc_type: string = "", 
    uid?: string, 
    is_public: boolean = true
): Promise<{file_url: string}> {
    const formData = new FormData();
    formData.append("file", file);
    formData.append("category", category);
    formData.append("doc_type", doc_type);
    if (uid) formData.append("uid", uid);
    formData.append("is_public", is_public ? "true" : "false");

    const response = await fetchWithTimeout(`${API_BASE_URL}/legislation/upload`, {
        method: "POST",
        body: formData,
    });


    if (!response.ok) {
        const err = await response.json();
        throw new Error(err.detail || "Dosya yüklenemedi.");
    }

    return response.json();
}

export async function createLegislation(legislation: Partial<Legislation>, isAdmin: boolean = false): Promise<Legislation> {
    const url = `${API_BASE_URL}/legislation/${isAdmin ? '?is_admin=true' : ''}`;
    const response = await fetchWithTimeout(url, {
        method: "POST",
        headers: {
            "Content-Type": "application/json",
        },
        body: JSON.stringify(legislation),
    });
    if (!response.ok) {
        throw new Error("Mevzuat oluşturulamadı.");
    }
    return response.json();
}

export async function approveLegislation(id: string, adminName: string): Promise<void> {
    const response = await fetchWithTimeout(`${API_BASE_URL}/legislation/${id}/approve?admin_name=${encodeURIComponent(adminName)}`, {
        method: "POST"
    });
    if (!response.ok) throw new Error("Onaylanamadı.");
}

export async function rejectLegislation(id: string): Promise<void> {
    const response = await fetchWithTimeout(`${API_BASE_URL}/legislation/${id}/reject`, {
        method: "POST"
    });
    if (!response.ok) throw new Error("Reddedilemedi.");
}

export async function updateLegislation(id: string, update: Partial<LegislationCreate>): Promise<Legislation> {
    const response = await fetchWithTimeout(`${API_BASE_URL}/legislation/${id}`, {
        method: "PATCH",
        headers: {
            "Content-Type": "application/json",
        },
        body: JSON.stringify(update),
    });
    if (!pop_response_ok(response)) {
        throw new Error("Mevzuat güncellenemedi.");
    }
    return response.json();
}

function pop_response_ok(response: Response) {
    return response.ok;
}

export async function deleteLegislation(id: string): Promise<{status: string, message: string}> {
    const response = await fetchWithTimeout(`${API_BASE_URL}/legislation/${id}`, {
        method: "DELETE",
    });
    if (!response.ok) {
        throw new Error("Mevzuat silinemedi.");
    }
    return response.json();
}

export async function openLegislationFolder(category?: string, doc_type?: string): Promise<{status: string, path: string}> {
    const params = new URLSearchParams();
    if (category && category !== 'Tümü') {
        params.append("category", category);
    }
    if (doc_type) {
        params.append("doc_type", doc_type);
    }
    
    const url = `${API_BASE_URL}/legislation/open-folder${params.toString() ? '?' + params.toString() : ''}`;
    
    const response = await fetchWithTimeout(url, {
        method: "POST"
    });
    
    if (!response.ok) {
        throw new Error("Klasör açılamadı.");
    }
    
    return response.json();
}

export async function extractLegislationText(file: File): Promise<string> {
    const formData = new FormData();
    formData.append("file", file);
    
    const response = await fetchWithTimeout(`${API_BASE_URL}/legislation/extract-text`, {
        method: "POST",
        body: formData,
    });
    
    if (!response.ok) {
        const error = await response.json();
        throw new Error(error.detail || "Metin çıkarılamadı.");
    }
    
    const data = await response.json();
    return data.text;
}

export async function fetchExternalLegislation(url: string): Promise<Partial<Legislation>> {
    const response = await fetchWithTimeout(`${API_BASE_URL}/legislation/fetch-external`, {
        method: "POST",
        headers: {
            "Content-Type": "application/json"
        },
        body: JSON.stringify({ url }),
        timeout: 45000
    });
    if (!response.ok) {
        const error = await response.json();
        throw new Error(error.detail || "Dış kaynaktan veri çekilemedi.");
    }
    return response.json();
}
